stream_alert/rule_processor/threat_intel.py

```diff
@@ -13,90 +13,389 @@
+        for record in records:
+            ioc_collections.extend(self._extract_values(record))
+
+        # Query DynamoDB IOC type to verify if the extracted info are malicious IOC(s)
+        self._process_ioc(ioc_collections)
+
+        # IOC info will be inserted to the records if they contains malicious IOC(s)
+        for ioc in ioc_collections:
+            if ioc.is_ioc:
+                self._insert_ioc_info(ioc.associated_record, ioc.ioc_type, ioc.value)
+
+        records_with_ioc = [ioc.associated_record for ioc in ioc_collections if ioc.is_ioc]
+        return records_with_ioc
+
+    def _insert_ioc_info(self, rec, ioc_type, ioc_value):
+        """Instance method to insert ioc info to a record.
+        It also removes ioc info duplication.
+
+        Args:
+            rec (dict): The parsed payload of a log
+            ioc_type (str): IOC type, can be 'ip', 'domain' or 'md5'
+            ioc_value (str): Malicious IOC value
+
+        Returns:
+            (None): A record will be modified in-place with IOC info inserted.
+            Example that a record will be inserted with new field "streamalert:ioc":
+                "streamalert:ioc": {
+                    "ip": [4.3.2.1", "1.2.3.4"],
+                    "domain" : ["evil1.com", "evil2.com"]
+                  }
+        """
+        if self.IOC_KEY in rec:
+            if (ioc_type in rec[self.IOC_KEY] and
+                    ioc_value not in rec[self.IOC_KEY][ioc_type]):
+                rec[self.IOC_KEY][ioc_type].append(ioc_value)
+            else:
+                rec[self.IOC_KEY][ioc_type] = [ioc_value]
+        else:
+            rec.update({self.IOC_KEY: {ioc_type: [ioc_value]}})
+
+    def _extract_values(self, record):
+        """Instance method to extract IOC info from the record based on normalized keys
+
+        Args:
+            record (dict): Normalized record.
+
+        Returns:
+            (list): Return a list of StreamIoc instances.
+        """
+        ioc_values = set()
+        for datatype in record[NORMALIZATION_KEY]:
+            # Lookup mapped IOC type based on normalized CEF type from Class variable.
+            ioc_type = self.__normalized_types2ioc_types.get(datatype, None)
+
+            # A new StreamIoc instance will be created when normalized CEF type
+            # has mapped IOC type.
+            if ioc_type:
+                for original_keys in record[NORMALIZATION_KEY][datatype]:
+                    value = record
+                    if isinstance(original_keys, list):
+                        for original_key in original_keys:
+                            value = value[original_key]
+                    ioc_values.add(value)
+        return [StreamIoc(value=value, ioc_type=ioc_type, rec=record) for value in ioc_values]
 
     @classmethod
-    def read_compressed_files(cls, intel_dir, delimiter=','):
-        """Read intelligence into memory
+    def load_config(cls, config):
```

Comment:
Can you make this method return an instance of the `StreamThreatIntel` class (or `None` if it's not enabled). This would allow you to pass the region, table, etc to the `__init__` instead of dynamically swapping out the class's protected properties (`__enabled`, `__region`, `__table`) in this method. Doing this is bad practice and should be avoided.